### PR TITLE
Modular PDAs: The End of an Era

### PR DESCRIPTION
**Pull request by JohnWildkins**
### TL;DR
PDAs are dead, long live PDAs. All trace of old PDAs has been scoured from the codebase, and in its place are modular computer PDAs that are feature-equivalent. Essentially every PDA function except the Syndicate detonation feature and Notepad has been ported over, and battery life for handheld computers has been boosted alongside the addition of charging cables to make things easier.

This is absolutely not bug-free, likely not 100% feature-complete, and definitely in need of (perhaps multiple) test merge(s).

### Changelog
- Programs
  - One new Vue UI: Janitorial Supplies Locator
  - Two programs made open-use: Atmosphere Scanner and Signaler
  - Three PDA functions added as PDA services: Gas, Medical, and Reagent Analyzer
  - NTIRC (messenger) now requires a registered ID to connect. IDs should be registered and NTIRC pre-loaded when you join the round, but if necessary you can simply restart NTIRC to auto-register, or register/unregister manually through hardware configuration. An ID can only be registered to one computer at a time.
- New Hardware
  - Charging Cable: Simply use a handheld computer with this installed on any APC to charge directly from the mains.
  - Hotswappable Battery: An otherwise standard battery that can be removed and installed without tools.
  - Signaler Card: A more power-hungry standard network card that provides signaler functions.
- Misc
  - Character Setup option added to select desired PDA type (tablet, wristbound, old-style PDA, or none)
  - ID card slots on modular computers now also store pens and similar-size items, as PDAs once did.
  - Lots of other QoL tweaks I've already forgotten about that bring modular computer functionality more in line with PDA functionality
  - and probably a bunch more that I forgot, it's been like two weeks of solid coding, okay

### IC changelog
`NanoTrasen is pleased to announce the development of their latest handheld operating system - NtOS version 2.1.0! In collaboration with the recently-acquired Thinktronic Solutions LTD, NanoTrasen engineers have integrated many of the adored quality of life features of Thinktronic's popular Personal Data Assistant line with the latest computing technology in their tablet and wristbound minicomputer lineup. A new line of Thinktronic PDAs have flooded the market alongside this development in celebration, boasting a native install of NtOS over their own homegrown solution for the first time in fifty years.`

### Optional to-do's
Not blocking by any means, but might be nice to do if I feel motivated to get them done before merge
- [ ] NTIRC (and other programs requiring network connection) should attempt to re-connect every thirty seconds or so, perhaps up to three times, rather than failing outright and requiring restart.
- [ ] Signaler cards, batteries, and cables should probably be sold somewhere rather than just buildable
- [ ] Probably make the MULEbot / Security bot control programs actually accessible

### CODE REVIEW

**Comment by JohnWildkins** [comment]
!review

**Review by Geevies — APPROVED** [review]

**Review by Chada-1 — APPROVED** [review]

File: code/modules/modular_computers/computers/subtypes/dev_pda.dm [hunk 1]
```diff
@@ -0,0 +1,14 @@
+/obj/item/modular_computer/handheld/pda
+	name = "Personal Data Assistant"
+	desc = "The latest in portable microcomputer solutions from Thinktronic Systems, LTD."
+	var/icon_add // this is the "bar" part in "pda-bar"
+	enrolled = DEVICE_PRIVATE
+
+/obj/item/modular_computer/handheld/pda/Initialize()
+	. = ..()
+	icon = 'icons/obj/pda.dmi'
+	icon_state = "pda"
```
Comment (NonQueueingMatt) [inline code/modules/modular_computers/computers/subtypes/dev_pda.dm:10]: why's this set here?

File: code/modules/modular_computers/computers/subtypes/preset_handheld.dm [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+/obj/item/modular_computer/handheld/preset/command
+	_app_preset_type = /datum/modular_computer_app_presets/command
+
+/obj/item/modular_computer/handheld/preset/command/Initialize()
+	. = ..()
+	card_slot.stored_item = new /obj/item/pen/fountain/head
+
+/obj/item/modular_computer/handheld/preset/command/cciaa
+	_app_preset_type = /datum/modular_computer_app_presets/command
+	// detonate = 0
```
Comment (NonQueueingMatt) [inline code/modules/modular_computers/computers/subtypes/preset_handheld.dm:143]: commented code all over this file, delete it

File: code/modules/modular_computers/file_system/programs/generic/ntnrc_client.dm [hunk 3]
```diff
@@ -230,26 +256,63 @@
 		ui_header = "ntnrc_idle.gif"
 
 /datum/computer_file/program/chatclient/kill_program(var/forced = FALSE)
+	if(!forced)
+		var/confirm = alert("Are you sure you want to close the NTNRC Client? You will not be reachable via messaging if you do so.", "Close?", "Yes", "No")
+		if(!(confirm == "Yes") || (CanUseTopic(usr) != STATUS_INTERACTIVE))
```
Comment (NonQueueingMatt) [inline code/modules/modular_computers/file_system/programs/generic/ntnrc_client.dm:261]:
```suggestion
		if(confirm != "Yes" || (CanUseTopic(usr) != STATUS_INTERACTIVE))
```

File: code/modules/modular_computers/file_system/programs/generic/signaler.dm [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+			return
+		var/mob/living/silicon/pai/host = true_computer.computer_host
+		return host.sradio
+	else
+		if(!computer || !computer.network_card?.sradio)
+			return
+
+		return computer.network_card.sradio
+
+// Gaters data for ui
```
Comment (NonQueueingMatt) [inline code/modules/modular_computers/file_system/programs/generic/signaler.dm:49]:
```suggestion
// Gathers data for ui
```

File: code/modules/modular_computers/file_system/programs/generic/signaler.dm [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	. = ..()
+
+	radio = get_radio()
+	if(!radio)
+		return
+
+	if(href_list["send"])
+		radio.send_signal("ACTIVATE")
+		for(var/mob/O in hearers(1, get_turf(computer)))
+			O.show_message(text("[icon2html(host, viewers(get_turf(src)))] *beep* *beep*"), 3, "*beep* *beep*", 2)
```
Comment (NonQueueingMatt) [inline code/modules/modular_computers/file_system/programs/generic/signaler.dm:82]: this should be a visible message

File: code/modules/modular_computers/file_system/programs/generic/signaler.dm [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+		if(new_frequency < PUBLIC_LOW_FREQ || new_frequency > PUBLIC_HIGH_FREQ)
+			new_frequency = sanitize_frequency(new_frequency)
+		radio.set_frequency(new_frequency)
+		return 1
+
+	else if(href_list["code"])
+		radio.code += href_list["code"]
+		radio.code = round(radio.code)
+		radio.code = min(100, radio.code)
+		radio.code = max(1, radio.code)
```
Comment (NonQueueingMatt) [inline code/modules/modular_computers/file_system/programs/generic/signaler.dm:96]: do `Clamp(1, round(radio.code), 100)`

File: code/modules/modular_computers/hardware/network_card.dm [hunk 7]
```diff
@@ -12,6 +12,7 @@ var/global/ntnet_card_uid = 1
 	var/identification_string = ""	// Identification string, technically nickname seen in the network. Can be set by user.
 	var/long_range = FALSE
 	var/ethernet = FALSE // Hard-wired, therefore always on, ignores NTNet wireless checks.
+	var/obj/item/radio/integrated/signal/sradio = FALSE // integrated signaler - not present on basic model.
```
Comment (NonQueueingMatt) [inline code/modules/modular_computers/hardware/network_card.dm:15]: delete this on Destroy

File: code/_helpers/unsorted.dm [hunk 8]
```diff
@@ -313,11 +313,12 @@ Turf and target are seperate in case you want to teleport some distance from a t
 					if(!search_pda)	break
 					search_id = 0
 
-			else if( search_pda && istype(A,/obj/item/device/pda) )
-				var/obj/item/device/pda/PDA = A
-				if(PDA.owner == oldname)
-					PDA.owner = newname
-					PDA.name = "PDA-[newname] ([PDA.ownjob])"
+			else if( search_pda && istype(A,/obj/item/modular_computer) )
```
Comment (NonQueueingMatt) [inline code/_helpers/unsorted.dm:316]:
```suggestion
			else if(search_pda && istype(A, /obj/item/modular_computer))
```

File: code/modules/modular_computers/computers/subtypes/preset_pda.dm [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+	icon_add = "c"
+	// detonate = 0
+
+/obj/item/modular_computer/handheld/pda/command/captain/Initialize()
+	. = ..()
+	card_slot.stored_item = new /obj/item/pen/fountain/captain
+
+/obj/item/modular_computer/handheld/pda/command/bst
+	icon_add = "h"
+	// hidden = 1
```
Comment (NonQueueingMatt) [inline code/modules/modular_computers/computers/subtypes/preset_pda.dm:200]: commented code all over here too

File: code/modules/modular_computers/hardware/tesla_link.dm [hunk 10]
```diff
@@ -5,9 +5,69 @@
...
+	var/obj/machinery/power/P = power_source
+	if(!istype(P))
+		return
+	source = P
+	tether(source)
+
+/obj/item/computer_hardware/tesla_link/charging_cable/proc/deactivate()
+	untether()
+	if(source)
+		source = null
```
Comment (mikomyazaki) [inline code/modules/modular_computers/hardware/tesla_link.dm:50]: This is already done in untether()

File: code/modules/modular_computers/hardware/tesla_link.dm [hunk 11]
```diff
@@ -5,9 +5,69 @@
...
+		activate(power_source)
+	else
+		to_chat(user, SPAN_NOTICE("You disconnect \the [src]."))
+		deactivate()
+
+/obj/item/computer_hardware/tesla_link/charging_cable/proc/activate(var/power_source)
+	var/obj/machinery/power/P = power_source
+	if(!istype(P))
+		return
+	source = P
```
Comment (mikomyazaki) [inline code/modules/modular_computers/hardware/tesla_link.dm:44]: This is already done in tether()

File: vueui/src/components/view/mcomputer/system/config.vue [hunk 12]
```diff
@@ -1,19 +1,24 @@
...
-    <i>Welcome to computer configuration utility. Please consult your system administrator if you have any questions about your device.</i><hr>
+    <i>Welcome to the NanoTrasen™ computer configuration utility.
+      Please consult your system administrator if you have any questions about your device.</i><hr>
     <h2>Power Supply</h2>
     <vui-item label="Battery Status:">
       <span v-if="battery">Active</span>
       <span v-else>Not Available</span>
     </vui-item>
-    <vui-item v-if="battery" label="Battery Rating:">{{battery.rating}}</vui-item>
+    <vui-item v-if="battery" label="Battery Rating:">{{battery.rating}} Wh</vui-item>
```
Comment (mikomyazaki) [inline vueui/src/components/view/mcomputer/system/config.vue:10]: Aren't these ratings in mWh according to your changes?

**Review by mikomyazaki — CHANGES_REQUESTED** [review]

File: code/modules/mob/abstract/new_player/preferences_setup.dm [hunk 13]
```diff
@@ -22,6 +22,7 @@
 	randomize_hair_color("facial")
 
 	backbag = 2
+	pda_choice = 2
```
[the remark this answers is not in the document]
Comment (JohnWildkins) [reply code/modules/mob/abstract/new_player/preferences_setup.dm:25]: i think it's sorta meant to be so that you can change what styles mean what if necessary without the database having a stinky

File: code/modules/mob/living/silicon/computer.dm [hunk 14]
```diff
@@ -57,7 +57,7 @@
 
 /mob/living/silicon/pai/proc/personal_computer_interact()
 	set name = "Access Local Computer"
-	set category = "Subsystems"
+	set category = "bruh"
```
[the remark this answers is not in the document]
Comment (JohnWildkins) [reply code/modules/mob/living/silicon/computer.dm:60]: bruh

File: vueui/src/components/view/mcomputer/system/config.vue (hunk 12 above)
Comment (JohnWildkins) [reply vueui/src/components/view/mcomputer/system/config.vue:10]: I think the standard is still that power_usage is watts, so since I'm changing it properly now to be per second, watt-hours is fine. I want to go through and do a proper overhaul of power at some point anyway to actually make units displayed uniformly everywhere, since it's a huge mess right now.

**Review by Snakebittenn — APPROVED** [review]

File: code/datums/outfits/outfit.dm [hunk 15]
```diff
@@ -272,16 +295,18 @@
...
-	PDA.name = "PDA-[H.real_name] ([PDA.ownjob])"
+/datum/outfit/proc/imprint_pda(mob/living/carbon/human/H, obj/item/modular_computer/PDA)
+	PDA.name = "PDA-[H.real_name] ([get_id_assignment(H)])"
+	PDA.update_icon()
+
+/datum/outfit/proc/register_pda(obj/item/modular_computer/P, obj/item/card/id/I)
+	P.card_slot.insert_id(I)
+	if(!P.hidden)
+		return
+	if(P.card_slot && P.card_slot.stored_card)
```
Comment (mikomyazaki) [inline code/datums/outfits/outfit.dm:306]:
```suggestion
	if(P.card_slot.stored_card)
```

If we do not have a card slot, we already runtimed on line 303. This check should be before line 303.

**Review by mikomyazaki — CHANGES_REQUESTED** [review]

**Review by mikomyazaki — APPROVED** [review]

**Comment by JohnWildkins** [comment]
merge this without testmerging and we will all die

File: code/modules/modular_computers/computers/modular_computer/core.dm [hunk 16]
```diff
@@ -93,6 +93,7 @@
 		install_default_programs()
 	handle_verbs()
 	update_icon()
+	initial_name = name
```
Comment (mikomyazaki) [inline code/modules/modular_computers/computers/modular_computer/core.dm:96]: Why do we store things in an initial_name var? Can't we just use initial(name)?

**Review by mikomyazaki — DISMISSED** [review]

File: code/modules/modular_computers/computers/modular_computer/core.dm (hunk 16 above)
Comment (mikomyazaki) [reply code/modules/modular_computers/computers/modular_computer/core.dm:96]: Nevermind. Apparently you can rename computers in-game.

**Review by mikomyazaki — APPROVED** [review]

File: code/modules/modular_computers/file_system/programs/generic/ntnrc_client.dm [hunk 17]
```diff
@@ -56,7 +56,8 @@
 		if(!channel)
 			return TRUE
 		var/mob/living/user = usr
-		user.visible_message("[SPAN_BOLD("\The [user]")] taps on [user.get_pronoun("his")] computer's screen.")
+		if(istype(user, /mob/living/carbon/human))
```
Comment (mikomyazaki) [inline code/modules/modular_computers/file_system/programs/generic/ntnrc_client.dm:59]:
```suggestion
		if(ishuman(user))
```

File: code/modules/modular_computers/file_system/programs/generic/ntnrc_client.dm [hunk 18]
```diff
@@ -69,7 +70,8 @@
 			to_chat(usr, SPAN_WARNING("The target chat isn't active on your program anymore!"))
 			return
 		var/mob/living/user = usr
-		user.visible_message("[SPAN_BOLD("\The [user]")] taps on [user.get_pronoun("his")] computer's screen.")
+		if(istype(user, /mob/living/carbon/human))
```
Comment (mikomyazaki) [inline code/modules/modular_computers/file_system/programs/generic/ntnrc_client.dm:73]:
```suggestion
		if(ishuman(user))
```

**Review by mikomyazaki — DISMISSED** [review]
